### PR TITLE
Pcs resource wait

### PR DESCRIPTION
**Pull request by atemp**
hi again,

here is a small hack that we use to wait for resources. i am not sure if it is a good generic way to do it like this, so if you dislike it, no problem. At least this time `--wait` is supported by 0.9 and 0.10.

I also read #28, so maybe `pcs_wait_for` would be a better solution, but it seems it is not here yet.

Rationale: We had problems that resources will take some time to get up and running after getting created and the next tasks in playbook/role then would fail.

First we tried to use `--wait` at the end of options string. That works fine when creating a service but it fails when the service is already here, e.g.:
```
TASK [el8cs-cluster : DLM resource] ********************************************************************************
fatal: [inf155]: FAILED! => {"changed": false, "error": "Error: Cannot use '-f' together with '--wait'\n", "msg": "Unable to simulate resource with given definition using command 'pcs -f /tmp/tmpnb77vb8b resource create dlm ocf:pacemaker:controld args=\"--protocol=sctp\" op monitor interval=30s on-fail=fence --group locking --wait'", "output": ""}
```

So what this pull request does is passes `--wait` only when creating the resource. Thus i called the option `wait_on_create`. Besides that we noticed, that if we add the `--wait` to the next run_command (the comment says it is only used in el6) it would fail, too. So it seems that in CentOS 8.1 this second cal is triggered and it is important to have no `--wait` there. We had no time to debug this yet.

To sum it up: I just wanted to show you what we did, if you think it is not useful for everyone, just reject this pull request. We will happily use `pcs_wait_for` once it is available.

### CODE REVIEW

**Comment by OndrejHome** [comment]
Hi @atemp ,
Thank you for nice summary on what you use and for adding into "we would like to use `pcs_wait_for` too once available". 

> Rationale: We had problems that resources will take some time to get up and running after getting created and the next tasks in playbook/role then would fail.

I like this idea and it is the point that was raised in #28 as the best solution as far as I remember.

I'm not very familiar with `--wait` option on `pcs` so I would have to have a look at it before deciding on what to do with PR you have proposed here. I'm wondering how efficient is this hack/trick in your deployment. Is it 100% of situations or more like something over XX% success?

As I can see the `dlm` resource which is not the fastest but also not the slowest I might guest if this just doesn't add a little extra time that would make it succeed most of the time. I mean that my worry is how this would do if there was resource that takes lets say more than 10 seconds to start, but may 1-2 seconds to create.

In any case I will have a better look at this and the older PR and will let you know on how I will proceed. I didn't managed to get time today, so I will check this tomorrow evening (my time).

Thank you for sharing!

**Comment by OndrejHome** [comment]
Having a look at `--wait` in CentOS 6.10, 7.7, 8.1 it looks that it is there really to wait for the resource to start according to documentation.
~~~
# pcs --version; cat /etc/centos-release 
0.9.155
CentOS release 6.10 (Final)

# pcs resource create --help
...
    create <resource id> [<standard>:[<provider>:]]<type> [resource options]
...
        If --wait is specified, pcs will wait up to 'n' seconds for the resource
        to start and then return 0 if the resource is started, or 1 if
        the resource has not yet started.  If 'n' is not specified it defaults
        to 60 minutes.
~~~

~~~
# pcs --version; cat /etc/centos-release 
0.9.167
CentOS Linux release 7.7.1908 (Core)
# pcs resource create --help
...
    create <resource id> [<standard>:[<provider>:]]<type> [resource options]
...
        default settings. If --wait is specified, pcs will wait up to 'n'
        seconds for the resource to start and then return 0 if the resource is
        started, or 1 if the resource has not yet started. If 'n' is not
        specified it defaults to 60 minutes.
~~~

~~~
# pcs --version; cat /etc/centos-release 
0.10.2
CentOS Linux release 8.1.1911 (Core)
# pcs resource create --help
...
    create <resource id> [<standard>:[<provider>:]]<type> [resource options]
...
        default settings. If --wait is specified, pcs will wait up to 'n'
        seconds for the resource to start and then return 0 if the resource is
        started, or 1 if the resource has not yet started. If 'n' is not
        specified it defaults to 60 minutes.
~~~

At this moment I think that this still makes sense as it is will be easier to implement than the `pcs_wait_for` which can be used also in more generic way during operation of cluster and not just resource creation. However I would check further what will be needed for this `wait_on_create` to get in. **Take a rest of this as me thinking aloud.**
First of concerns is that it only can be enabled/disable which is 'no wait' vs '60 minutes of waiting'. Maybe having attribute that will default to '0' meaning no '--wait' with ability to give it number which will result in '--wait number' would be nice. 
Because of `'pcs %(cib_file_param)s resource create...` part there would need to be check that will remove it when `cib_file` parameter is used - which is valid use case for people building more complex configuration first into file and the applying it in cluster. Possibly we could just drop it and print warning in such case.
~~~
            if rc != 0 and "Call cib_replace failed (-62): Timer expired" in err:
                # EL6: special retry when we failed to create resource because of timer waiting on cib expired
~~~
I would be quite surprised if we are hitting this in CentOS 8, but I haven't got many deployments on CentOS 8 as the packages for cluster got there only recently. On EL6 I have suspected this to be present because of something "being slow" after the cluster was first created and ansible being "too fast" hitting it with resource creation. I will see if I can reproduce this, but just in case that you manage to hit this again I would like to check the logs from the time when this happened - specifically `/var/log/pacemaker/pacemaker.log` from CentOS 8.1 should contain the details on what was attempted to be writtent to CIB when this error came.

For now, don't worry that this has conflict with master branch (mostly because of documentation updates). I need to let this sink for a bit for me and after that I will make a plan how to move forward with this. I have some free days this week, so it shouldn't take much. Thanks for your patience and for bringing this up.

**Comment by atemp** [comment]
> First of concerns is that it only can be enabled/disable which is 'no wait' vs '60 minutes of waiting'. Maybe having attribute that will default to '0' meaning no '--wait' with ability to give it number which will result in '--wait number' would be nice.

I thought of the same thing, but in that case i suggest to use a second argument for the time, e.g. `wait_time` since i consider to specify e.g. `-1` for default looks ugly to me. I did not want to polish it yet, before i know that you like it.

> I would be quite surprised if we are hitting this in CentOS 8, but I haven't got many deployments on CentOS 8 as the packages for cluster got there only recently.

I can try to reproduce that and open an issue if you want me to. I have some livesys (lorax/livemedia-creator) based cluster setup now, so i can easily rollout test clusters, as long as they do no thave to reboot (or fence) and come back again.

CentOS 8.1 has HA but no RS, so dlm is missing, if you want a quick fix for that feel free to take my rebuilds from source here: http://repos.tilab.tuwien.ac.at/public/el8/x86_64/emp-centos81-rs (highly unofficial, but works for me)

**Comment by OndrejHome** [comment]
> I can try to reproduce that and open an issue if you want me to.

Sounds good. Lets move this into Issue so this PR can stay for the original purpose of `wait_on_create` - I have created Issue #32 with minimal background, lets continue discussing this part there please.

> I thought of the same thing, but in that case i suggest to use a second argument for the time, e.g. wait_time since i consider to specify e.g. -1 for default looks ugly to me. I did not want to polish it yet, before i know that you like it.

`-1` also looks quite ugly to me, but I think that `0` should be an usable number indicating "don't specify `--wait number`", any other positive number resulting in `--wait number`. Also good idea to get the plan first before getting to code. As mentioned, I will give it a bit of thinking and summarize here what would make sense to me to go forward. Lets discuss it once it I prepare it. Thanks for the inputs.

**Comment by OndrejHome** [comment]
Hi @atemp ,

I have made a summary on what I would expect from this option to get this merged:
- name: `wait_on_create`
- type: int
- default: 0
  - when '0' don't add it to any commands (current behaviour)
  - when '>0' add '--wait=XX' to `pcs resource create` commands (except ones with `-f`)
- ERROR when used together with `cib_file` -- check somewhere in the beggining for this
- print WARNING when `wait_on_create` is used but we are modifying or deleting resource 
   - WARNING: `wait_on_create` option is ignored when modifying or deleting the resource.
- when creation of resource fails with exit code 1 check if ti failed because of Timer or syntax error
  - if there was 'Error: waiting timeout' text in output lets error with message like "Timed out waiting for resource to start after XXs. Consider increasing 'wait_on_create' parameter."
  - in any other case the print the current error message "Failed to create resource using command '...'"

I have also checked that behaviour is consistent between CentOS 6.10, 7.7, 8.1 when using the `--wait` option and below outputs can be used for reference and testing.
~~~
[root@fastvm-centos-6-10-161 ~]# time pcs resource create test ocf:pacemaker:Dummy op_sleep=5 --wait=10; echo $?
Error: waiting timeout
Resource 'test' is not running on any node

Pending actions:
        Action 3: test_monitor_10000    on fastvm-centos-6-10-161
        Action 2: test_start_0  on fastvm-centos-6-10-161
Error performing operation: Timer expired

real    0m10.289s
user    0m0.181s
sys     0m0.054s
1
===
[root@fastvm-centos-6-10-161 ~]# time pcs resource create test ocf:pacemaker:Dummy op_sleep=5 --wait=20; echo $?
Resource 'test' is running on node fastvm-centos-6-10-161.

real    0m16.300s
user    0m0.179s
sys     0m0.064s
0
~~~ 
~~~ 
[root@fastvm-centos-7-7-171 ~]# time pcs resource create test ocf:pacemaker:Dummy op_sleep=5 --wait=10; echo $?
Error: waiting timeout

Pending actions:
        Action 3: test_monitor_10000    on fastvm-centos-7-7-171
Error performing operation: Timer expired

real    0m10.262s
user    0m0.167s
sys     0m0.070s
1
===
[root@fastvm-centos-7-7-171 ~]# time pcs resource create test ocf:pacemaker:Dummy op_sleep=5 --wait=20; echo $?
resource 'test' is running on node 'fastvm-centos-7-7-171'

real    0m16.282s
user    0m0.170s
sys     0m0.086s
0
~~~ 
~~~ 
[root@fastvm-centos-8-1-181 ~]# time pcs resource create test ocf:pacemaker:Dummy op_sleep=5 --wait=10; echo $?
Error: waiting timeout

Pending actions:
        Action 3: test_monitor_10000    on fastvm-centos-8-1-181
Error performing operation: Timer expired

real    0m10.440s
user    0m0.331s
sys     0m0.085s
1
===
[root@fastvm-centos-8-1-181 ~]# time pcs resource create test ocf:pacemaker:Dummy op_sleep=5 --wait=20; echo $?
resource 'test' is running on node 'fastvm-centos-8-1-181'

real    0m16.479s
user    0m0.368s
sys     0m0.082s
0
~~~ 

Please let me know what you think about above proposal and also if you think that it can be used with your use case/scenario. You can also choose if you would like to (a) adjust this PR to above list or (b) if you would prefer me to implement this - I'm OK with any approach you choose as the overall idea makes sense to me to be part of the module. If you choose to implement this ('a' option) please just let me know rough timeline (like "next week"/"next month") when you would have time to look at this (no need to hurry) so I can check with you if there are any difficulties.

In any case Thanks again for bringing this idea up. Contribution appreciated :+1:

**Comment by OndrejHome** [comment]
Hi @atemp,
In the meantime I had a progress on implementing the `crm_wait_for` module that should be covering also the use case described in this PR. You can use it to block progress of playbook until the resource gets into desired state. Additionally it is possible to configure how often the state is checked and also set timeout after which the module will fail if the desired state was not reached.

- [Github: ansible.crm_wait_for](https://github.com/OndrejHome/ansible.crm_wait_for)
- [Ansible galaxy: ondrejhome.crm_wait_for](https://galaxy.ansible.com/OndrejHome/crm_wait_for)

I will be therefore closing this PR, but feel free to re-open this PR or open Issue against the `crm_wait_for` module in case there is something missing for your use case. 
Thank you!
